### PR TITLE
Fetch valid RSEs and keep it up-to-date in MSPileup

### LINKED ISSUE #11498: Setup proper mandatory/optional values for MSPileup document
**Impact of the bug**
We got initial feedback from data-ops about MSPileup services and Alan summarized mandatory, optional and service attributes in MSPileup requirement [document](https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Pileup). We should incorporate them into MSPileupObj.py module.

**Describe the bug**
Current implementation of MSPileupObj.py have current defaults for some of the mandatory attributes, and we should revisit all defaults for optional ones and ensure proper values for services attributes.

**How to reproduce it**
Read [MSPileupObj.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py) code for data assingment.

**Expected behavior**
Have proper set of mandatory, optional and service attributes according to requirement [document](https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Pileup).

**Additional context and error message**

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #11498 (sort of!)

#### Status
In development

#### Description
This PR provides the following:
* reads a configurable rudio Disk expression `rucioDiskExpression` to be used to find valid and allowed Rucio RSEs from Rucio (which relies on a memory cache set to expiry every 12h, by default)
* it enforces RSE validation when creating a new pileup document
* when user updates a pileup object, validation of the parameters will be performed as well
* when MSPileupTasks updates a pileup object, validation of the document is skipped (given that it's already loaded from the database)

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

### CODE REVIEW

**Comment by amaltaro** [comment]
@vkuznet Valentin, could you please have a look at this initial implementation for the RSE validation? If it's looking Okay for you, I can update unit tests as well.

**Review by vkuznet — CHANGES_REQUESTED** [review]
Alan, I think you incorrectly provide the logic to MSPileup, while in my opinion it should go to MSPIleupTaskManager (i.e. the update action should be done by task manager rather HTTP server). This is HTTP service, as such it only provides HTTP methods like GET/POST/DELETE/PUT. From your your changes I see multiple flaws:
- you changed the behavior of pileup thread which should only retrieve the status
- unless you want to provide HTTP way to modify validRSEs as an external request this implementation does not belong to MSPileup and should go to MSPileupTaskManager and executed as independent task
- MSPileup class is rucio free since it does not involved in any data placement operation

**Comment by amaltaro** [comment]
@vkuznet Valentin, as we discussed over Zoom, RSEs need to be validated between the HTTP request and the action to persist it in the database, so this MSPileup service is the proper place to have the definition of the valid RSEs on (instead of the Tasks instance).

I updated this PR as follows:
* 2nd commit provides a fix on how we actually use MemoryCache in the Rucio wrapper and ensuring it won't grow when we are not using the cache
* 3rd commit provides a lock to the MSPileupData instance variable `validRSEs`, to avoid possible race conditions between reading and updating the list

The standard libraries don't seem to provide any way to perform separate read/write locks (blocking reads only based on the write lock). For that, we would have to implement a custom class, [example](https://gist.github.com/tylerneylon/a7ff6017b7a1f9a506cf75aa23eacfd6) (but I didn't check the full implementation).

I am still not very happy with the use of this lock and maybe the best would be to fetch RSEs from the Rucio wrapper cache for every createPileup/updatePileup call and pass that over for the validation.

**Comment by amaltaro** [comment]
And with the 4th commit - which is very much incomplete - I show how we could get rid of the threading lock mechanism. In plain english, using the Rucio wrapper cache, we could fetch the list of RSEs whenever there is an incoming Create or Update pileup object call.

Note the placeholder for the updatePileup method, which turns out not to enforce any user data validation at the current moment.

**Comment by vkuznet** [comment]
> And with the 4th commit - which is very much incomplete - I show how we could get rid of the threading lock mechanism. In plain english, using the Rucio wrapper cache, we could fetch the list of RSEs whenever there is an incoming Create or Update pileup object call.
> 
> Note the placeholder for the updatePileup method, which turns out not to enforce any user data validation at the current moment.

So, you are gradually converging to what I suggested, having rucio call for every HTTP request and passing around rses. Yes, in this implementation we do not need locks and it will be free from racing conditions.

Meanwhile you'll do your changes, please separate `updateRSE()` and `status()` calls. The former updates RSE, the latter get the status, therefore the former should not include the latter. I suggest to remove status call from updateRSE and use them separately in pileup function in MSManager.

**Comment by amaltaro** [comment]
> So, you are gradually converging to what I suggested, having rucio call for every HTTP request and passing around rses. Yes, in this implementation we do not need locks and it will be free from racing conditions.

Okay, I take it that the 4th commit provides an approach that you like the better as well. 
I will modify the code appropriately, but it might be that I won't get it ready by the end of the day tomorrow. Thanks Valentin.

**Comment by amaltaro** [comment]
@vkuznet Valentin, I still have to check the jenkins report and fix the unit tests. But if you have any important feedback on the latest implementation, please let me know.

**Comment by amaltaro** [comment]
@vkuznet Valentin, there are a couple of MSPileupTasks unit tests that now fail and the reason is that now document update is validated as well, while the current master code does not run any validation on top of the user-provided data (which in our case could be coming from MSPileupTasks as well).

So I would like to ask your opinion on the following potential implementations and/or behaviors:
* first question is whether we should always validate data being updated to the documents. My opinion is that end-users must have everything updated; while data provided by the microservice itself don't need to go through the whole MSPileupObj validation
* it looks like we read the document from MongoDB twice as well (one too many!). First when executing the MSPileupTasks tasks; second when calling MSPileupData.updatePileup.
* to avoid the double read and thorough document validation for microservice-originated data, we could create a "private" method in MSPileup that will be used only by MSPileupTasks.

Please let me know what your thoughts are. Others are always welcomed to chime in as well, of course!

**Comment by vkuznet** [comment]
In my opinion, the validation should be done in all steps, regardless if user create or update the document. Lack of validation in update step may lead to silly mistakes from end-user, e.g. a typo, which will provide invalid docs at the end.

For other two items you asked for I need more context. But I think you are talking about the following block https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSPileup/MSPileupData.py#L130-L143 If so, I think it would be much cleaner to remove it to avoid double read and let code fail in update if document does not exist (if you'll modify the code you may change `insert` to `update` in failed message on line 151.

**Comment by amaltaro** [comment]
> In my opinion, the validation should be done in all steps, regardless if user create or update the document. Lack of validation in update step may lead to silly mistakes from end-user, e.g. a typo, which will provide invalid docs at the end.

Yes, we agree on that!

However, there are two scenarios where a pileup document can be updated:
1) from user-provided arguments, through the RESTful API.
2) from microservice-provided arguments, in this case MSPileupTasks, which loads every single document from the database, change it if needed and persist the up-to-date document back in the database.

So, my question is specific to the case 2). After MSPileupTasks performed all the actions on a given MSPileupObj object and it's about to make a call to `updatePileup` to get the updated document persisted in the database, do you think we should pass the new dictionary through the MSPileupObj.validate call, such that we verify the data before persisting it? Or would you say that we should be safe (enough), assuming that the microservice is doing the right thing, thus simply by-passing any further MSPileupObj.validate call?

If you think we should not validate it, then I believe we would need to create a "private" method, e.g. `_updatePileup`, which will be used only by MSPileupTasks, such that the data structure passed to be persisted isn't validated against the MSPileupObj class.

If I was not clear enough, maybe a 10min Zoom chat will be better.

**Comment by vkuznet** [comment]
Alan, I still do not understand why do you need any private method, why you can't just add additional parameter to updatePileup to skip validation and make it obligatory by default, e.g. `def updatePileup(self, pdict, validate=True)`. This way there is no need for another private method, and you can clearly distinguish that all HTTP clients will by default use validation and MSPileupTasks can provide this extra parameter to skip it.

**Comment by amaltaro** [comment]
I think I am done with changes to this PR and it's ready for a full review.

File: src/python/WMCore/MicroService/MSPileup/MSPileup.py [hunk 1]
```diff
@@ -80,28 +80,30 @@ def queryDatabase(self, query, projection=None):
...
         """
-        MSPileup update API to update correspoding pileup document in data layer.
+        MSPileup update API to update corresponding pileup document in data layer.
 
         :param pdict: input MSPilup data dictionary
         :return: results of MSPileup data layer (list of dicts)
         """
         self.authMgr.authorizeApiAccess('ms-pileup', 'update')
-        return self.mgr.updatePileup(pdict)
+        rseNames = self.rucio.evaluateRSEExpression(self.diskRSEExpr, useCache=True)
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSPileup/MSPileup.py:105]: Alan, I thought we agree to not use the cache, while here it is the opposite. Could you please elaborate on this `useCache=True` option.

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py [hunk 2]
```diff
@@ -113,37 +114,58 @@ def createPileup(self, pdict):
             return [err.error()]
         return []
 
-    def updatePileup(self, doc):
+    def updatePileup(self, doc, rseList=None, validate=False):
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSPileup/MSPileupData.py:117]: I thought that default should always be `validate=True` and only upstream code will explicitly change this parameter.

File: src/python/WMQuality/Emulators/EmulatedUnitTestCase.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def setUp(self):
 
         TODO: parameters to turn off emulators individually
         """
-
+        print(f"Running setUp with emulated services. ")
```
Comment (vkuznet) [inline src/python/WMQuality/Emulators/EmulatedUnitTestCase.py:48]: why print and not logger call?

File: test/data/Mock/RucioMockData.json [hunk 4]
```diff
@@ -1404,6 +1404,7 @@
   "account": "ms-pileup"
  }],
  "evaluateRSEExpression": ["T2_XX_CERN"],
+ "evaluateRSEExpression:[('useCache', True)]": ["rse1", "rse2"],
```
Comment (vkuznet) [inline test/data/Mock/RucioMockData.json:1407]: should we use useCache=False instead?

**Review by vkuznet — CHANGES_REQUESTED** [review]

File: src/python/WMCore/MicroService/MSPileup/MSPileup.py (hunk 1 above)
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSPileup/MSPileup.py:105]: No, we actually agreed to use the Rucio wrapper cache, specially because this call is tied to EVERY HTTP call, so we better use it not to increase latency in the HTTP calls (also because the RSE expression is always the same!).

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py (hunk 2 above)
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSPileup/MSPileupData.py:117]: Just for convenience, given that with this default, there is only 1 place calling it that needs to be changed (MSPileup). While the inverted logic would require changes in a handful of calls from MSPileupTasks (and maybe other places). Other than that, I have no strong preference and I could revert the logic if you want. Just let me know.

File: src/python/WMQuality/Emulators/EmulatedUnitTestCase.py (hunk 3 above)
Comment (amaltaro) [reply src/python/WMQuality/Emulators/EmulatedUnitTestCase.py:48]: No strong preference as well, but for many tests the logger buffer gets many many records, so a line like that is harder to see than if we get it in the stdout. I am fine to change it though, if you prefer.

File: test/data/Mock/RucioMockData.json (hunk 4 above)
Comment (amaltaro) [reply test/data/Mock/RucioMockData.json:1407]: We should use whatever signature of the call we have in place. Given that we are using the cache, it has to be True here as well.

File: src/python/WMCore/MicroService/MSPileup/MSPileup.py (hunk 1 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileup.py:105]: The discussion was about effect of the cache to HTTP service which should be simple. Said that, I will not argue here until we know that cache really significantly contribute to RAM usage. Let's proceed with `usageCache=True` and keep eye on RAM usage.

File: src/python/WMCore/MicroService/MSPileup/MSPileupData.py (hunk 2 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupData.py:117]: Regardless how much changes should be done, the default should be right here. I strongly suggest to make `validate=True` as default and make explicit calls when this should be skipped.

File: src/python/WMQuality/Emulators/EmulatedUnitTestCase.py (hunk 3 above)
Comment (vkuznet) [reply src/python/WMQuality/Emulators/EmulatedUnitTestCase.py:48]: The strange thing that you always made similar argument when I used print in a code. I followed your suggestion here :) since if codebase use logger everywhere it seems logical to use it here too.

File: test/data/Mock/RucioMockData.json (hunk 4 above)
Comment (vkuznet) [reply test/data/Mock/RucioMockData.json:1407]: fine.

File: src/python/WMQuality/Emulators/EmulatedUnitTestCase.py (hunk 3 above)
Comment (amaltaro) [reply src/python/WMQuality/Emulators/EmulatedUnitTestCase.py:48]: I think for unit tests it gives us a good separation from the actual WMCore code, thus stdout would have only unit test related log records.

File: src/python/WMCore/MicroService/MSPileup/MSPileup.py (hunk 1 above)
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSPileup/MSPileup.py:105]: I don't think we need to be concerned with this because cache will be a dictionary with a single key/value pair. Any updates to the cache will be void because the key will already exist.

**Comment by amaltaro** [comment]
@vkuznet I think I covered most of your concerns, please have another look at your convenience.

**Review by vkuznet — APPROVED** [review]
Alan, changes look good to me, please proceed with merge.

**Comment by amaltaro** [comment]
Thank you very much for the review over the weekend, Valentin.

**Comment by amaltaro** [comment]
Tag 2.2.0rc1 should have these changes, plus MSAuth, plus fix for monitor.sh. Build pipeline is still running:
https://github.com/dmwm/WMCore/actions/runs/4398024756

Once we confirm CRUD operations are functional, I think we can then have another iteration of P&R tests.
